## Import Stuff

In [1]:
# Importing different Python libraries and packages
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Importing and saving the dataset for this problem set
df = pd.read_csv("problem_set_1_data.csv")
df

quizzes  attendedClass  finalgrade  final_EXAM              student
0      78.50              1       99.38          46         Linnie Lietz
1      64.25              0       95.78          45        Mitch Mustain
2      76.75              0       97.42          41       Salina Chavera
3      69.00              1       92.50          38    Kimberely Conwell
4      64.00              0       92.50          39            Zack Burk
..       ...            ...         ...         ...                  ...
166    73.50              0       96.25          45           Dana Malta
167    75.50              0       98.13          44         Sixta Heyden
168    41.00              0       69.38          32       Brianne Broome
169    77.00              0      101.25          47  Elouise Weatherholt
170    72.00              0       92.81          38        Addie Maharaj

[171 rows x 5 columns]

## DataFrame Manipulation and Summary

In [9]:
# Rename the dataframe columns using the rename function in Pandas
df = df.rename(columns = {'quizzes':'quiz_total', 'attendedClass' : 'attended', 'finalgrade': 'final_grade', 'final_EXAM': 'final_exam', 'student': 'name'})

# Printing the head, or first 5, and tail, or last 5, of the dataframe information
print('----------------------------------------')
print("First 5 rows of the DataFrame")
print(df.head())

print('----------------------------------------')
print("Last 5 rows of the DataFrame")
print(df.tail())

# Summary of the Dataframe Numeric Information
print('----------------------------------------')
print("Description and Summary Statistics of the DataFrame")
print(df.describe())


----------------------------------------
First 5 rows of the DataFrame
   quiz_total  attended  final_grade  final_exam               name
0       78.50         1        99.38          46       Linnie Lietz
1       64.25         0        95.78          45      Mitch Mustain
2       76.75         0        97.42          41     Salina Chavera
3       69.00         1        92.50          38  Kimberely Conwell
4       64.00         0        92.50          39          Zack Burk
----------------------------------------
Last 5 rows of the DataFrame
     quiz_total  attended  final_grade  final_exam                 name
166        73.5         0        96.25          45           Dana Malta
167        75.5         0        98.13          44         Sixta Heyden
168        41.0         0        69.38          32       Brianne Broome
169        77.0         0       101.25          47  Elouise Weatherholt
170        72.0         0        92.81          38        Addie Maharaj
-------------------

- For the "quiz_total" column, the average quiz score was 69.13%
- The number of 0.192982 for those who attended could represent the percentage of students who attended the class schronously on Zoom (19.3%)
- The average final grade that the class as a whole received was around 91.86% which is an A-
- The average score that each student received on their final exam was around 39.68 out of 50 potentially

## Effects of Class Attendance on Grades

In [10]:
# Basing the Pandas DataFrame on those who attended class
attend_df = df[df['attended'] == 1]
print('----------------------------------------')
print('The averages of the final grade, final exam, and quiz total based on students that attended class')
print(attend_df.groupby('attended').mean())

# Basing the Pandas DtaFrame on those who did not attend class
absent_df = df[df['attended'] == 0]
print('----------------------------------------')
print('The averages of the final grade, final exam, and quiz total based on students that did not attend class')
print(absent_df.groupby('attended').mean())

        

----------------------------------------
The averages of the final grade, final exam, and quiz total based on students that attended class
          quiz_total  final_grade  final_exam
attended                                     
1               72.5    95.072121   40.636364
----------------------------------------
The averages of the final grade, final exam, and quiz total based on students that did not attend class
          quiz_total  final_grade  final_exam
attended                                     
0          68.326087    91.086159   39.442029


- The overall result of class attendence on overall grades conclude that students who attended class more frequently had higher average grades compared to absent students in the following: final grade (95.07% vs 91.09%), quiz total (72.5% vs 68.33%), and final exam grade (40.64% vs 39.44%)

## Bonus Points - Who Cheated?

In [8]:
def find_cheaters():
    duplicated_list = df.duplicated(subset = ['quiz_total', 'final_grade', 'final_exam'], keep = False)
    name_list = list(df['name'])
    
    new_df = pd.DataFrame({'Name': name_list, 'Duplicates': duplicated_list})
    new_dict = dict(sorted(new_df.values.tolist()))
#     print(new_dict)
    
    cheaters = []
    
    for key, value in new_dict.items():
        if value == False:
            pass
        else:
            names = key
            cheaters.append(names)
            
    return cheaters

cheaters = find_cheaters()
print(cheaters)

['Addie\xa0Maharaj', 'Jaraham Eidda']


- The cheaters are Addie Maharaj and Jaraham Eidda